cli/lib/kontena/cli/helpers/exec_helper.rb

```diff
@@ -1,87 +1,190 @@
+    #
+    # @param ws [Kontena::Websocket::Client]
+    # @param tty [Boolean]
+    # @return [Thread]
+    def websocket_exec_write_thread(ws, tty: nil)
+      Thread.new do
+        begin
+          read_stdin(tty: tty) do |stdin|
+            websocket_exec_write(ws, 'stdin' => stdin)
+          end
+          websocket_exec_write(ws, 'stdin' => nil) # EOF
+        rescue => exc
+          logger.error exc
+          ws.close(1001, "stdin read #{exc.class}: #{exc}")
+        end
+      end
     end
 
-    # @param container_id [String] The container id
+    # Connect to server websocket, send from stdin, and write out messages
+    #
+    # @param paths [String]
+    # @param options [Hash] @see Kontena::Websocket::Client
+    # @param cmd [Array<String>] command to execute
     # @param interactive [Boolean] Interactive TTY on/off
     # @param shell [Boolean] Shell on/of
     # @param tty [Boolean] TTY on/of
-    # @return [String]
-    def ws_url(container_id, interactive: false, shell: false, tty: false)
-      require 'uri' unless Object.const_defined?(:URI)
-      extend Kontena::Cli::Common unless self.respond_to?(:require_current_master)
+    # @return [Integer] exit code
+    def websocket_exec(path, cmd, interactive: false, shell: false, tty: false)
+      exit_status = nil
+      write_thread = nil
 
-      url = URI.parse(require_current_master.url)
-      url.scheme = url.scheme.sub('http', 'ws')
-      url.path = "/v1/containers/#{container_id}/exec"
-      if shell || interactive || tty
-        query = {}
-        query.merge!(interactive: true) if interactive
-        query.merge!(shell: true) if shell
-        query.merge!(tty: true) if tty
-        url.query = URI.encode_www_form(query)
-      end
-      url.to_s
-    end
+      query = {}
+      query[:interactive] = interactive if interactive
+      query[:shell] = shell if shell
+      query[:tty] = tty if tty
 
-    # @param [String] url
-    # @param [String] token
-    # @return [WebSocket::Client::Simple]
-    def connect(url, token)
-      options = {
-        headers: {
+      server = require_current_master
+      url = websocket_url(path, query)
+      token = require_token
+      options = WEBSOCKET_CLIENT_OPTIONS.dup
+      options[:headers] = {
           'Authorization' => "Bearer #{token.access_token}"
-        }
       }
-      if ENV['SSL_IGNORE_ERRORS'].to_s == 'true'
-        options[:verify_mode] = ::OpenSSL::SSL::VERIFY_NONE
+      options[:ssl_params] = {
+          verify_mode: ENV['SSL_IGNORE_ERRORS'].to_s == 'true' ? OpenSSL::SSL::VERIFY_NONE : OpenSSL::SSL::VERIFY_PEER,
+          ca_file: server.ssl_cert_path,
+      }
+      options[:ssl_hostname] = server.ssl_subject_cn
+
+      logger.debug { "websocket exec connect... #{url}" }
+
+      # we do not expect CloseError, because the server will send an 'exit' message first,
+      # and we return before seeing the close frame
+      # TODO: handle HTTP 404 errors
```

Comment:
The old code had some special handling of HTTP 404 responses, which this is missing. ATM something like a `kontena container exec core-01/weavex` looks like this:

```
 [error] Error during WebSocket handshake: Unexpected response code: 404
```

Compared to other commands:

```
 [error] 404 : Not found (/v1/services/development/null/test)
```

Probably acceptable... otherwise the `kontena-websocket-client` would need to trap the `ProtocolError` raised by the `websocket-driver` and reraise it as some kind of `HTTPError` with a `@status_code`.